File: cpp/inheritance.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <memory>

using namespace std;

// An abstract (base) class to model a logging interface with two APIs.
class LoggerInterface {
  public:
    virtual ~LoggerInterface() = 0;
    virtual void Log(string const& mesg) = 0;
    virtual vector<string> Messages() const = 0;
};

LoggerInterface::~LoggerInterface() {}


// InMemoryLogger saves the log messages in memory.
class InMemoryLogger : public LoggerInterface {
  public:
    InMemoryLogger() = default;
    virtual ~InMemoryLogger() = default;
    virtual void Log(string const& mesg);
    virtual vector<string> Messages() const;
  private:
    vector<string> messages;
};

void InMemoryLogger::Log(string const& mesg) {
  messages.push_back(mesg);
}

vector<string> InMemoryLogger::Messages() const {
  return messages;
}


// LocalLogger saves the log messages in a file.
class LocalLogger : public LoggerInterface {
  public:
    LocalLogger(char const * const file);
    virtual ~LocalLogger();
    virtual void Log(string const& mesg);
    virtual vector<string> Messages() const;
  private:
    char const * const filename{nullptr};
    fstream outfile;
};

LocalLogger::LocalLogger(char const * const file) :
  filename{file} {
  outfile.open(filename, fstream::out);
  if (!outfile.is_open()) {
    cerr << "Failed to open file: " << filename << '\n';
    exit(1);
  }
}

void LocalLogger::Log(string const& mesg) {
  outfile << mesg << '\n';
  outfile.flush();
}

vector<string> LocalLogger::Messages() const {
  vector<string> messages;
  ifstream infile(filename);
  for (string line; getline(infile, line); ) {
    messages.push_back(line);
  }
  return messages;
}

LocalLogger::~LocalLogger() {
  outfile.close();
}


// Helper function to emit vector<string>.
ostream& operator<<(ostream& os, vector<string> const& vs) {
  for (auto const& s : vs) {
    os << s << '\n';
  }
  return os;
}


int main(int argc, char * argv[1]) {
  char const * const default_filename = "/tmp/outfile_cpp.txt";
  char const * const filename = (argc > 1) ? argv[1] : default_filename;

  // A sequential collection of interfaces via vector of unique pointers.
  vector<unique_ptr<LoggerInterface>> loggers;
  loggers.emplace_back(make_unique<InMemoryLogger>());
  loggers.emplace_back(make_unique<LocalLogger>(filename));

  vector<string> const testMessages = {
    "Hello, World!",
    "abracadabra",
    "Sayonara!"
  };
  for (auto const& mesg : testMessages) {
    for (auto& logger : loggers) {
      logger->Log(mesg);
    }
  }

  for (auto& logger : loggers) {
    auto const observedMessages = logger->Messages();
    if (observedMessages != testMessages) {
      cerr << "expected: " << testMessages
           << "; but observed: " << observedMessages << '\n';
      exit(1);
    }
  }
}

File: golang/inheritance.go
package main

import (
  "bufio"
  "fmt"
  "log"
  "os"
  "reflect"
)

// A logging interface with two APIs.
type LoggerInterface interface {
  Log(string)
  Messages() []string
}

func Log(li LoggerInterface, mesg string) {
  li.Log(mesg)
}

func Messages(li LoggerInterface) []string {
  return li.Messages()
}


// InMemoryLogger saves the log messages in memory.
type InMemoryLogger struct {
  LoggerInterface
  messages []string
}

func (this *InMemoryLogger) Log(mesg string) {
  this.messages = append(this.messages, mesg)
}

func (this *InMemoryLogger) Messages() []string {
  return this.messages
}


// LocalLogger saves the log messages in a file.
type LocalLogger struct {
  LoggerInterface
  filename string
  file *os.File
}

func makeLocalLogger(filename string) *LocalLogger {
  file, err := os.OpenFile(filename, os.O_RDWR|os.O_CREATE, 0755)
  if err != nil {
    log.Fatal(err)
  }
  return &LocalLogger{filename: filename, file:file}
}

func (this *LocalLogger) Log(mesg string) {
  fmt.Fprintln(this.file, mesg)
}

func (this *LocalLogger) Messages() []string {
  file, err := os.Open(this.filename)
  if err != nil {
    log.Fatal(err)
  }
  defer file.Close()

  var messages []string
  scanner := bufio.NewScanner(file)
  for scanner.Scan() {
    messages = append(messages, scanner.Text())
  }
  if err := scanner.Err(); err != nil {
    log.Fatal(err)
  }
  return messages
}


func main() {
  filename := "/tmp/outfile_golang.txt";
  if len(os.Args) > 1 {
    filename = os.Args[1]
  }

  // A sequential collection of interfaces via slices.
  var loggers []LoggerInterface
  loggers = append(loggers, &InMemoryLogger{})
  loggers = append(loggers, makeLocalLogger(filename))

  var testMessages = []string{
    "Hello, World!",
    "abracadabra",
    "Sayonara!",
  }
  for _, mesg := range(testMessages) {
    for _, logger := range(loggers) {
      logger.Log(mesg)
    }
  }

  for _, logger := range(loggers) {
    observedMessages := logger.Messages()
    if !reflect.DeepEqual(observedMessages, testMessages) {
      log.Fatal("expected: ", testMessages,
        "; but observed: ", observedMessages, "\n")
    }
  }
}
